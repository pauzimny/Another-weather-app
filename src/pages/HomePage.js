import React, { useState, useEffect } from "react";
import Result from "../components/Result.js";
import Form from "../components/Form.js";
import Message from "../components/Message";

const HomePage = props => {
  const APIKEY = "1ea9d8ecd83d43c9a0801ba96a6b4ae1";
  let tempArray = [];
  let initialResults = JSON.parse(localStorage.getItem("cities")) || [];

  const [error, setError] = useState(false);
  const [isAlready, setIsAlready] = useState(false);
  const [units, setUnits] = useState("metric");
  const [input, setInput] = useState("");
  const [results, setResults] = useState([...initialResults]);
  const [getData, setGetData] = useState(false);
  const [msg, setMsg] = useState();

  useEffect(() => {
    if (props.location.state !== undefined) {
      setUnits(props.location.state.units);
    }
    if (getData) {
      const index = results.findIndex(
        city => city.name.toLowerCase() === input.toLowerCase()
      );
      if (index === -1) {
        const API = `https://api.openweathermap.org/data/2.5/forecast?q=${input}&lang=pl&units=${units}&appid=${APIKEY}`;

        console.log(input);
        fetch(API)
          .then(response => {
            if (response.ok) {
              return response.json();
            }
            setError(true);
            setMsg(
              `Miasto nie istnieje w bazie lub nie mozna pobrać danych dla miasta ${input}`
            );
            throw Error(`Nie mozna pobrać danych dla miasta ${input}`);
          })
          .then(data => {
            data.list.forEach(measure => {
              tempArray.push(measure.main.temp);
            });
            const sumTemp = tempArray.reduce((a, b) => a + b, 0);
            const averageTemp = sumTemp / tempArray.length;
            const city = {
              name: input.toLocaleLowerCase(),
              id: input,
              temp: Math.round(averageTemp),
              units,
              lon: data.city.coord.lon,
              lat: data.city.coord.lat
            };
            setResults([...results, city]);
            setInput("");
            setGetData(false);
          })
          .catch(() => {
            setInput("");
          });
      } else {
        setIsAlready(true);
        setMsg("To miasto jest juz na liście");
        setInput("");
        setError(false);
        setGetData(false);
      }
    }
    localStorage.setItem("cities", JSON.stringify(results));
  }, [getData]);

  const handleChange = e => {
    setInput(e.target.value);
    setIsAlready(false);
    setError(false);
    setGetData(false);
    setMsg("");
  };

  const handleSubmit = e => {
    e.preventDefault();
    setGetData(true);
  };

  const handleDelete = id => {
    localStorage.clear();
    const cities = [...results];
    const index = cities.findIndex(city => city.id === id);
    cities.splice(index, 1);
    setResults([...cities]);
    window.localStorage.setItem("cities", JSON.stringify(cities));
  };

  const citiesResults = results.map(result => (
    <Result
      id={result.id}
      key={result.name}
      number={result.id}
      city={result.name}
      temp={result.temp}
      delete={handleDelete}
      lat={result.lat}
      lon={result.lon}
      units={result.units}
      err={result.err}
    />
  ));
  return (
    <div>
      <Form submit={handleSubmit} change={handleChange} value={input} />
      {error && <Message msg={msg} />}
      {isAlready && <Message msg={msg} />}
      <ol className="resultList">{citiesResults}</ol>
    </div>
  );
};

export default HomePage;
