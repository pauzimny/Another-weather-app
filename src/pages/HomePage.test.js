import HomePage from "./HomePage.js";
import React from "react";
import { shallow, render, mount } from "enzyme";
import toJson from "enzyme-to-json";
import sinon from "sinon";

it("renders correctly", () => {
  const wrapper = shallow(
    <HomePage
      submit={() => {}}
      change={() => {}}
      value={"london"}
      error={false}
      isAlready={false}
      citiesResults={[
        {
          id: "london",
          key: "london",
          number: "london",
          city: "london",
          temp: 12,
          delete: () => {},
          lat: 15,
          lon: 15,
          units: "metric",
          err: false
        }
      ]}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it("renders correctly again", () => {
  const wrapper = render(
    <HomePage
      submit={() => {}}
      change={() => {}}
      value={"london"}
      error={false}
      isAlready={false}
      citiesResults={[
        {
          id: "london",
          key: "london",
          number: "london",
          city: "london",
          temp: 12,
          delete: () => {},
          lat: 15,
          lon: 15,
          units: "metric",
          err: false
        }
      ]}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

// it("calls handleSubmit on submit", () => {
//   const spy = sinon.spy();
//   const wrapper = mount(
//     <HomePage
//       submit={spy}
//       change={() => {}}
//       value={"london"}
//       error={false}
//       isAlready={false}
//       citiesResults={[
//         {
//           id: "london",
//           key: "london",
//           number: "london",
//           city: "london",
//           temp: 12,
//           delete: () => {},
//           lat: 15,
//           lon: 15,
//           units: "metric",
//           err: false
//         }
//       ]}
//     />
//   );

//   wrapper
//     .find("div")
//     .first()
//     .simulate("submit");

//   expect(spy.calledOnce).toBe(true);
// });
