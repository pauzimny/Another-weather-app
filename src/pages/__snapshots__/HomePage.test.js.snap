// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="home-page"
>
  <Form
    change={[Function]}
    submit={[Function]}
    value=""
  />
  <ResultHeader />
  <ol
    className="resultList"
  />
</div>
`;

exports[`renders correctly again 1`] = `
<div
  class="home-page"
>
  <form
    class="form"
  >
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-textField-1 MuiFormControl-marginNormal"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="standard-with-placeholder"
        style="color:#ff680c"
      >
        Nazwa miasta
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-input-2 MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="standard-with-placeholder"
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-addButton-3 MuiButton-contained"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Dodaj miasto
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          style="margin-left:5px"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </span>
    </button>
  </form>
  <div
    class="MuiPaper-root MuiPaper-elevation8 makeStyles-resultHeader-86 MuiPaper-rounded"
  >
    <span
      class="result-header__num"
    >
      #
    </span>
    <span
      class="result-header__city"
    >
      Nazwa miasta
    </span>
    <span
      class="result-header__temp"
    >
      Åšrednia prognozowana temperatura
    </span>
    <span
      class="result-header__delete"
    />
  </div>
  <ol
    class="resultList"
  />
</div>
`;
